appcues/src/main/java/com/appcues/data/remote/adapters/TraitResponseAdapterFactory.kt

```diff
@@ -0,0 +1,66 @@
+package com.appcues.data.remote.adapters
+
+import com.appcues.data.remote.response.trait.TraitResponse
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonDataException
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import java.lang.reflect.Type
+
+class TraitResponseAdapterFactory : JsonAdapter.Factory {
+
+    override fun create(type: Type, annotations: MutableSet<out Annotation>, moshi: Moshi): JsonAdapter<*>? {
+        return if (
+            Types.getRawType(type) == List::class.java &&
+            Types.collectionElementType(type, List::class.java) == TraitResponse::class.java
+        ) {
+            val delegate = moshi.adapter(TraitResponse::class.java)
+            TraitResponseAdapter(delegate)
+        } else {
+            null
+        }
+    }
+
+    private class TraitResponseAdapter(
+        private val delegate: JsonAdapter<TraitResponse>,
+    ) : JsonAdapter<List<TraitResponse>>() {
+
+        override fun fromJson(reader: JsonReader): List<TraitResponse> {
+            val traits = mutableListOf<TraitResponse>()
+            val traitTypes = mutableSetOf<String>()
+            val duplicates = mutableSetOf<String>()
+
+            reader.beginArray()
+
+            while (reader.hasNext()) {
+                val value = reader.readJsonValue()
+                delegate.fromJsonValue(value)?.let {
+                    if (!traitTypes.add(it.type)) {
+                        // If adding the type to the existing set returns false, this means it
+                        // was already in the set, and this is a duplicate type in this collection.
+                        // Duplicates are collected and used in a decoding error.
+                        duplicates.add(it.type)
+                    } else {
+                        // Normal case, capture the decoded trait for our resulting list
+                        traits.add(it)
+                    }
+                }
+            }
+
+            if (duplicates.isNotEmpty()) {
+                val message = "multiple traits of same type are not supported: ${duplicates.joinToString(",")}. " +
```

Comment:
So the way this works - the set below is valid, since they are different trait types. They would be 3 different types of backdrop decorating traits perhaps.
```
"traits": [
  {
    "type": "backdrop"
  },
  {
    "type": "backdrop-1"
  },
  {
    "type": "backdrop-2"
  }
]
```

This other set below would NOT be valid - there are two `backdrop-2` traits - two of the same type at the same level, which we do not allow.
```
"traits": [
  {
    "type": "backdrop-1"
  },
  {
    "type": "backdrop-1"
  },
  {
    "type": "backdrop-2"
  }
]
```

The message would indicate which trait is the issue: "multiple traits of same type are not supported: backdrop-1. Found at path xxxxx.xxx.xxx"

The logic here loops over every trait in the JSON, and finds ALL such instances where a certain type has been duplicated (not just the first one) and then sends an error message if there are any dupes, containing all of those dupe trait types.

Reply:
Traits, in JSON, do only have the `type`, no other name, so that's all we have to go on for the de-dupe.

Reply:
oh yea my bad.